arch/risc-v/include/irq.h

```diff
@@ -64,12 +64,70 @@ extern "C"
 #define EXTERN extern
 #endif
 
+/****************************************************************************
+ * Inline Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_irq_save
+ *
+ * Description:
+ *   Disable interrupts and return the previous value of the mstatus register
+ *
+ ****************************************************************************/
+
+static inline irqstate_t up_irq_save(void)
+{
+  irqstate_t flags;
+
+  /* Read mstatus & clear machine interrupt enable (MIE) in mstatus */
+
+  __asm__ __volatile__
+    (
+      "csrrc %0, mstatus, %1\n"
+      : "=r" (flags)
+      : "r"(MSTATUS_MIE)
+      : "memory"
+    );
+
+  /* Return the previous mstatus value so that it can be restored with
+   * up_irq_restore().
+   */
+
+  return flags;
+}
+
+/****************************************************************************
+ * Name: up_irq_restore
+ *
+ * Description:
+ *   Restore the value of the mstatus register
+ *
+ ****************************************************************************/
+
+static inline void up_irq_restore(irqstate_t flags)
+{
+  __asm__ __volatile__
+    (
+      "csrw mstatus, %0\n"
+      : /* no output */
+      : "r" (flags)
+      : "memory"
+    );
+}
+
 /****************************************************************************
  * Public Function Prototypes
  ****************************************************************************/
 
-EXTERN irqstate_t  up_irq_save(void);
-EXTERN void up_irq_restore(irqstate_t);
+/****************************************************************************
+ * Name: up_irq_enable
+ *
+ * Description:
+ *   Return the current interrupt state and enable interrupts
+ *
+ ****************************************************************************/
+
 EXTERN irqstate_t up_irq_enable(void);
```

Comment:
should we inline this function too?

Reply:
I tried this implement, but found ESP32-C3 can't do `asm volatile("csrrs %0, mie, %1" : "=r"(oldstat) : "r"(MIE_MEIE));` this will cause a exception of `Illegal instruction`, so I have to keep this implement in each chip.

Reply:
But up_irq_enable should modify MSTATUS register not MIE. up_irq_[save|restore|enable] should operate on the global interrupt enable/disable bit(e.g. MSTATUS.MIE), not the individual irq line(MIE). The later API is up_enable_irq/up_disable_irq

Reply:
Yes, `up_irq_enable` is enable global interrupt of RISC-V CPU, but some chips should also enable MIE_MEIE which is used to enable machine external interrupt in this function.
If we inline this function too, we should put asm volatile("csrrs %0, mie, %1" : "=r"(oldstat) : "r"(MIE_MEIE)) into other place.